### PR TITLE
Consolidate docker containers

### PR DESCRIPTION
**Pull request by rs--**
Update the Docker containers the CI team supports:
- Consolidate the oetools `full`, `minimal`, and `deploy` containers to a single, `oetools-18.04/20.04` container (supports the functionality of all three).
- Generate Windows Server 2019 container

### CODE REVIEW

**Comment by rs--** [comment]
Waiting on full E2E run before merging. The containers produced from this branch were validated via the test build process on CDPx.

**Review by CyanDevs — APPROVED** [review]
This is a well done and thorough PR. Awesome work!

Some very minor thoughts:
* Could use a space after some commas in the Jenkinsfile parameters
* Could speed up builds a bit by copying only necessary files in oe, rather than the entire folder. Though it's not worth the effort if files are scattered everywhere.

**Comment by rs--** [comment]
Tests passed, and verified the new containers.

File: .jenkins/infrastructure/build_azure_managed_images.Jenkinsfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/build_azure_managed_images.Jenkinsfile:29]: I would recommend adding quotes around here along with brackets. "${params.IMAGE_VERSION}"

File: .jenkins/infrastructure/docker/build_linux_docker_images.Jenkinsfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/build_linux_docker_images.Jenkinsfile:7]: You do not want a default here, is that correct?

File: .jenkins/infrastructure/docker/build_linux_docker_images.Jenkinsfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/build_linux_docker_images.Jenkinsfile:25]: Can we further parallelize here so that it builds all containers at the same time on different hosts?

File: .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile:8]: /nit I would personally change the '0.8' - > 'x.x'

File: scripts/ansible/roles/linux/openenclave/vars/ubuntu/main.yml
Comment (BRMcLaren) [inline scripts/ansible/roles/linux/openenclave/vars/ubuntu/main.yml:14]: @anakrish suggested we keep clang-format-7. I'll let him comment on the reason, but I believe you may want to revert line 14 here.

File: .jenkins/infrastructure/docker/dockerfiles/windows/Dockerfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/dockerfiles/windows/Dockerfile:1]: I see no issue with keeping the docker windows images reliant on the PowerShell set up script. I would recommend a minor  readme.md here to explain the decision for future developers.

File: .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile:81]: We no longer use terraform, we can remove the installation here (unless you are working with it).

File: .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile:87]: We no longer use oe-engine either, as Azure Confidential COmputing has gone into public preview with their DC skus and we are using those.

File: .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile:79]: unzip and apt-transport-https are not Microsoft packages to my knowledge, I think they can be moved to line 70

File: .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile:37]: Do you think we should default to latest here?

Might be good to add a boolean check here. The reasoning is it would simplify debugging for devs if there are image issues so we do not override production images.

File: .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile:36]: I think there is a bracket issue here.

Line 36 should be on line 35, and 35 on 36.

File: .jenkins/infrastructure/build_docker_images.Jenkinsfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/build_docker_images.Jenkinsfile:25]: Would you mind if we did

"Linux Docker Container Build" -> "Linux-Docker-Container-Build" and for windows as well?

File: scripts/ansible/roles/linux/openenclave/vars/ubuntu/main.yml
Comment (BRMcLaren) [inline scripts/ansible/roles/linux/openenclave/vars/ubuntu/main.yml:38]: Is python required for this PR? Or is it specifically for the 20.04 support?

File: .jenkins/infrastructure/docker/build_linux_docker_images.Jenkinsfile
Comment (rs--) [reply .jenkins/infrastructure/docker/build_linux_docker_images.Jenkinsfile:7]: Yeah, I wanted the agent to be explicitly defined - we have defaults in Jenkins, so this shouldn't be an issue.

File: .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile
Comment (rs--) [reply .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile:36]: I think this is the intended behaviour we want; based on the Linux container build it seems like we only want to update the Docker Hub registry if `TAG_LATEST` is true. We do update the internal docker repository in line 27, with or without that flag.

File: .jenkins/infrastructure/build_docker_images.Jenkinsfile
Comment (rs--) [reply .jenkins/infrastructure/build_docker_images.Jenkinsfile:25]: Sure thing; updating the Jenkins job names now

File: scripts/ansible/roles/linux/openenclave/vars/ubuntu/main.yml
Comment (rs--) [reply scripts/ansible/roles/linux/openenclave/vars/ubuntu/main.yml:38]: It's specifically for 20.04; python3 is installed by default, but python isn't. It shouldn't affect 18.04, since it already has python installed.

File: .jenkins/infrastructure/docker/build_linux_docker_images.Jenkinsfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/build_linux_docker_images.Jenkinsfile:49]: .

File: .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile
Comment (BRMcLaren) [reply .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile:36]: It may be intended but is it best or is it done this way for a legacy reasons? 

In a year we may not be the ones maintaining this and having debuggable containers that are pullable outside of Azure infrastructure I believe to be preferable to simplify debugging. I'll leave the decision to you, but I would prefer to see pushing tagged containers regardless of latest every time and then latest dependent on the conditional.

File: .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/dockerfiles/linux/Dockerfile:76]: /nit

Technically you are installing only Azure CLI and not Microsoft packages :)

File: .jenkins/infrastructure/docker/dockerfiles/windows/Dockerfile
Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/dockerfiles/windows/Dockerfile:12]: >C:\oe


This is showing up in Github's highlighter. Could you check that it is correct and a character escape is not needed?

Comment (BRMcLaren) [inline .jenkins/infrastructure/docker/dockerfiles/windows/Dockerfile:12]: >oe/scripts/install-windows-prereqs.ps1

I believe Windows prefers to use a forward slash '\' instead of a backslash for standardization reasons. Mind changing this?

**Comment by BRMcLaren** [comment]
A couple of nits and comments, nothing blocking and will leave up to you if you want to implement.

Approving now, thanks for this work and the great addition!

**Review by BRMcLaren — APPROVED** [review]

File: .jenkins/infrastructure/docker/dockerfiles/windows/Dockerfile
Comment (rs--) [reply .jenkins/infrastructure/docker/dockerfiles/windows/Dockerfile:12]: Double checked, escape character isn't needed

File: .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile
Comment (rs--) [reply .jenkins/infrastructure/docker/build_windows_docker_images.Jenkinsfile:36]: I'll give the tags a more thorough look, and I'll address it in a future PR.

**Pull request merged**
